cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,529 @@
+        return doesLeaderReplicaRemoveViolateOptimizedGoal(replicaToBeRemoved) ? REPLICA_REJECT : ACCEPT;
+      case INTER_BROKER_REPLICA_SWAP:
+        Replica srcReplicaToSwap = clusterModel.broker(action.sourceBrokerId()).replica(action.topicPartition());
+        Replica dstReplicaToSwap = clusterModel.broker(action.destinationBrokerId()).replica(action.destinationTopicPartition());
+        return acceptReplicaSwap(srcReplicaToSwap, dstReplicaToSwap);
+      default:
+        throw new IllegalArgumentException("Unsupported balancing action " + action.balancingAction() + " is provided.");
+    }
+  }
+
+  private ActionAcceptance acceptReplicaSwap(Replica srcReplicaToSwap, Replica dstReplicaToSwap) {
+    if (srcReplicaToSwap.isLeader() == dstReplicaToSwap.isLeader()) {
+      return ACCEPT;
+    }
+    if (srcReplicaToSwap.isLeader()) {
+      return doesLeaderReplicaRemoveViolateOptimizedGoal(srcReplicaToSwap) ? REPLICA_REJECT : ACCEPT;
+    } else { // dstReplicaToSwap.isLeader()
+      return doesLeaderReplicaRemoveViolateOptimizedGoal(dstReplicaToSwap) ? REPLICA_REJECT : ACCEPT;
+    }
+  }
+
+  /**
+   * Check whether this goal is violated after removing the given leader replica. It is violated if there is not be enough
+   * leader replicas on that broker for that topic
+   *
+   * @param replicaToBeRemoved replica to be removed
+   * @return {@code true} if the given replica move would violate the optimized goal (i.e. the move is not acceptable),
+   * {@code false} otherwise.
+   */
+  private boolean doesLeaderReplicaRemoveViolateOptimizedGoal(Replica replicaToBeRemoved) {
+    if (!replicaToBeRemoved.isLeader()) {
+      // Moving a follower replica does not violate/affect this goal
+      return false;
+    }
+    int topicLeaderReplicaCountOnSourceBroker =
+        replicaToBeRemoved.broker().numLeaderReplicasOfTopicInBroker(replicaToBeRemoved.topicPartition().topic());
+    return topicLeaderReplicaCountOnSourceBroker <= minTopicLeadersPerBroker();
+  }
+
+  /**
+   * This is a hard goal; hence, the proposals are not limited to dead broker replicas in case of self-healing.
+   * Sanity Check: The total number of leader replicas of certain topics is sufficient so that these leader replicas
+   * can be moved to satisfy this goal.
+   *
+   * @param clusterModel The state of the cluster.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void initGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    if (_balancingConstraint.topicsWithMinLeadersPerBrokerPattern().pattern().isEmpty()) {
+      _mustHaveLeaderReplicaPerBrokerTopics = Collections.emptySet();
+      return;
+    }
+    _mustHaveLeaderReplicaPerBrokerTopics = Collections.unmodifiableSet(
+        Utils.getTopicNamesMatchedWithPattern(_balancingConstraint.topicsWithMinLeadersPerBrokerPattern(), clusterModel.topics()));
```

Comment:
`Utils.getTopicNamesMatchedWithPattern(...)` will also perform the empty pattern check and return an empty set. Can we eliminate repeated check for empty pattern?

Reply:
The `empty pattern check` in this `getTopicNamesMatchedWithPattern` method is intentional. In case the caller pass in an empty pattern and a long list of topic name. This empty check will catch it.

Reply:
I don't recommend a change in `Utils.getTopicNamesMatchedWithPattern(...)`. I refer to the preceding if clause.

Reply:
If we remove the empty check before `Utils.getTopicNamesMatchedWithPattern(...)`, this `clusterModel.topics()` method will be called and that will be a waste right?

Reply:
I see -- so the reason why there is a duplicate check is to avoid a relatively more expensive call to `clusterModel.topics()` in case the pattern is empty.
Thanks for the clarification.

Reply:
Actually we can achieve this optimization w/o code duplication by using a `Supplier<Set<String>>` as the second parameter of `Utils.getTopicNamesMatchedWithPattern(...)`.
Lazy evaluation ensures that the call for `clusterModel.topics()` be made only when needed.

Reply:
Good idea. Fixed